base/expr.jl

```diff
@@ -1207,19 +1254,28 @@ function make_atomic(order, ex)
 end
 function make_atomic(order, a1, op, a2)
     @nospecialize
-    is_expr(a1, :., 2) || error("@atomic modify expression missing field access")
-    a1l, a1r, op, a2 = esc(a1.args[1]), esc(a1.args[2]), esc(op), esc(a2)
-    return :(modifyproperty!($a1l, $a1r, $op, $a2, $order))
+    if is_expr(a1, :., 2)
+        a1l, a1r, op, a2 = esc(a1.args[1]), esc(a1.args[2]), esc(op), esc(a2)
+        return :(modifyproperty!($a1l, $a1r, $op, $a2, $order))
+    elseif is_expr(a1, :ref, 2)
+        x, idx, op, a2 = esc(a1.args[1]), esc(a1.args[2]), esc(op), esc(a2)
+        return :(modifyindex!($x, $idx, $op, $a2, $order))
```

Comment:
I think `idx` should ideally be supporting any combination of args/kwargs here, requiring rearranging the order of all these arguments though:
```suggestion
    elseif is_expr(a1, :ref)
        x, idxs, op, a2 = esc(a1.args[1]), map(esc, a1.args[2:end]), esc(op), esc(a2)
        return :(modifyindex_atomic!($x, $order, $op, $a2, $(idxs...)))
```

Reply:
I can change the order of arguments, but I'm not sure what `z = @atomic x[1:3]` would mean, as the whole operation will not be atomic (only the individual accesses to `x` will ?)

Or do we want to have this support in macro "for future use"?

Reply:
that won't work. the part that could is  `z = @atomic x[1, 3]`

Reply:
still, what I asked is how atomic `@atomic x[1,3]` is, if after `x[1]` is read (atomically) `x[3]` might have been modified by another thread before it is fetched to `z[2]`?

Reply:
that's not what `x[1,3]` does. this is a matrix indexing that pulls a single entry of the matrix out atomicslly

Reply:
`x[[1,3]]` would error

Reply:
so we want to support `GenericMemory{T} <: AbstractArray{T,1}` with 2-dimensional indexing? Or is is about already adding support for future `Memory{T,2} <: GenericMemory{T} <: AbstractArray{T,1}`?

I'm sorry but it seems that I'm very confused ;)

Reply:
I think the idea is that the macro should just work for a hypothetical user package (xref [FixedSizeArrays.jl](https://github.com/JuliaArrays/FixedSizeArrays.jl)) that would add methods to `getindex_atomic` and `setindex_atomic!`.

NB: FixedSizeArrays is a nonregistered package that will presumably support `GenericMemory`, although it doesn't yet. But see one of the open PRs.

Reply:
@vtjnash I've implemented the vararg support in the macro call